packages/babel-types/src/definitions/core.js

```diff
@@ -626,31 +647,54 @@ defineType("ObjectProperty", {
         };
       })(),
     },
-    value: {
-      validate: assertNodeType("Expression", "Pattern", "RestElement"),
-    },
+    value: {},
     shorthand: {
-      validate: assertValueType("boolean"),
+      validate: chain(
+        assertValueType("boolean"),
+        function(node, key, val) {
+          if (val && node.computed) {
+            throw new TypeError(
+              "Property shorthand of ObjectProperty cannot be true if computed is true",
+            );
+          }
+        },
+        function(node, key, val) {
+          if (val && !t.isIdentifier(node.key)) {
+            throw new TypeError(
+              "Property shorthand of ObjectProperty cannot be true if key is not an Identifier",
+            );
+          }
+        },
+      ),
       default: false,
     },
     decorators: {
+      optional: true,
       validate: chain(
         assertValueType("array"),
         assertEach(assertNodeType("Decorator")),
       ),
-      optional: true,
     },
   },
   visitor: ["key", "value", "decorators"],
   aliases: ["UserWhitespacable", "Property", "ObjectMember"],
+  validate: (function() {
+    const pattern = assertNodeType("Identifier", "Pattern");
+    const expression = assertNodeType("Expression");
+
+    return function(parent, key, node) {
+      const validator = t.isObjectPattern(parent) ? pattern : expression;
+      validator(node, "value", node.value);
+    };
+  })(),
 });
 
 defineType("RestElement", {
+  builder: ["argument"],
   visitor: ["argument", "typeAnnotation"],
-  aliases: ["LVal"],
   fields: {
     argument: {
-      validate: assertNodeType("LVal"),
+      validate: assertNodeType("Identifier"),
```

Comment:
It can also be a member expression, if the destructuring is not part of a parameter/declaration:
```js
var a = {};
[...a.b] = [1, 2, 3];
a.b; // 1, 2, 3
```